### ISSUE TITLE
[GPHOTOS:AUTH] Token is expired

### ISSUE THREAD

**Issue by Sachin1968** [hermanho/MMM-GooglePhotos#102]
After reinstalling MM and GooglePhotos everything worked fine for a couple of weeks, and stopped displaying anything. I reinstalled the module but erroring out at generate token step.

Log shows the following: [GPHOTOS:AUTH] Token is expired

I deleted previous tokens and downloaded a new credentials.json file but keep getting the error above. Is this a Google issue or something else?

Thanks,
Sachin

**Comment by aneaville** [comment]
After downloading/renaming the credentials.json, were you able to successfully run generate_token.js?

if you got 'token is expired', it makes me think that the token.json wasn't deleted.

**Comment by Sachin1968** [comment]
That was it, thank you. Is the Token expiring expected? Looks like it only lasted for a couple of weeks. In my previous install, the token was valid for a couple of years.

**Comment by aneaville** [comment]
The 'token' only lasts 1 hour. It refreshes when it does expire. 
Their is a separate 'refresh' token that lasts years(not sure how long) 
Sometimes the refreshing fails and can't get a new one. I don't know why it fails to try to fix it. It seems to fail after many days.

Lets leave this issue open and see if others have the token expiring issue after the new update.

**Comment by mtroxx** [comment]
I have had the same issue.  If I leave the system off for two days I have to delete the token.json file and then generate another token.  Is it possible to have a script delete the token.json file to make it easier to regenerate?  I would like to give a system to my sister but she'll never go in and manually delete the token.json.

**Comment by aneaville** [comment]
Is this a new issue with the latest changes(couple months ago), or has this been an ongoing thing?

**Comment by mtroxx** [comment]
It is happening to me now.  I built my first Magic Mirror (without the Mirror) about 3 months ago.  I have been working on setting one up for my sister.

**Comment by ed0zanki** [comment]
same here ... I am glad that I am not alone. Was about to go nuts .. :)
 would be nice if we had a sloution for that.

**Comment by ed0zanki** [comment]
is there anyone with a sloution or maybe anyone has any advice on what we are maybe doing wrong ?

**Comment by mtroxx** [comment]
I went into the directory using the gui interface and deleted the token.json file.  The system then prompted for the permissions - including the link to google.  This works for my situation.  Hope it helps yours.

**Comment by heitpi** [comment]
Even after deleting the token.json file and re-authenticating, this problem occurrs for me about every 7-10 days.  I have a daily reboot in cron, but this didn't solve the problem.  

@aneaville Regarding "Is this a new issue with the latest changes(couple months ago), or has this been an ongoing thing?"
  -- This was happening prior to the latest changes (it was a problem last year). My google photos albums are very large (~1000 photos), so I wonder if this is an instability related to the album size.

Is anyone else still experiencing this issue?

**Comment by Sachin1968** [comment]
> Even after deleting the token.json file and re-authenticating, this problem occurrs for me about every 7-10 days. I have a daily reboot in cron, but this didn't solve the problem.
> 
> @aneaville Regarding "Is this a new issue with the latest changes(couple months ago), or has this been an ongoing thing?"
> -- This was happening prior to the latest changes (it was a problem last year). My google photos albums are very large (~1000 photos), so I wonder if this is an instability related to the album size.
> 
> Is anyone else still experiencing this issue?

This is a Google token expiry issue. Mine worked for a few years then expired. The Google token policy now sets expiry in a few days. Read the beginning of my thread for details.

**Comment by aneaville** [comment]
I'll admit the whole Oauth 2.0 flow is not my expertise, but my understanding is that a refresh token should be good for a long time.

I see a couple possibilities:
1. According to Google's policy: https://developers.google.com/identity/protocols/oauth2 ,towards the very bottom:  "Google Cloud Platform project with an OAuth consent screen configured for an external user type and a publishing status of "Testing" is issued a refresh token expiring in 7 days."
   Maybe this app is listed as 'testing' and by changing that, the token would be good for longer.
Anyone know how to check this?

2. I think it may be possible that under some weird scenario, that the way the app gets a new token breaks and doesn't save the newest token. I think this would leave the app unable to continue to get new tokens. A point against this is heitpi rebooting daily and still seeing this issue.

There are google API's for Oath2. I'm not sure if authorization part of this app was based on one of those libraries or built separately. Its possible that switching to the Google API could improve things, but I'm not sure.

**Comment by aneaville** [comment]
I'm on a quest to get this 7-day expiring thing figured out. 

The first thing I'm going to try is changing my google cloud platform project 'publishing status'. Currently is is set to 'Testing', but changing to 'In Production' only took 1 click. No verification form or anything. 
I wasn't expecting this to be so easy, as my research made me think verification was the only way and was going to require I prove website ownership as the first step. But stumbling across this setting gives me hope. My guess is that for this 'platform project', I qualify under the 'personal use' exception. Even though I'm using open source code, the google console project is only for me. So maybe getting the app verified isn't necessary. This setting change seems like a simple change that could be added to the setup instructions. 

If you would like to try the same, go to your google cloud console, select your magic mirror project. Then from the navigation menu(top left) -> APIs & Services -> Oath consent screen. This should get you to a site something like https://console.cloud.google.com/apis/credentials/consent?project=[PROJECT_ID] where [PROJECT_ID] is the project ID. Here is where I found the publishing status. It looks like this:
![image](https://user-images.githubusercontent.com/13342606/112637617-57e04780-8e0c-11eb-86ad-cb65d305a389.png)



I've made the change and re-authorized. I'm not convinced this will fix the issue yet, but I'll post back here either way in a little over 7 days. There is a chance I might make some more changes, which may restart the 7 day timer, so if anyone else wants to try this as well, please share your results. 

-Andrew

**Comment by heitpi** [comment]
Hi Andrew,
Based on your post from weeks ago and recommendation to get out of "testing" mode, I tried this.  As you say, it is easy and is just a button click to "publish app".  I still had to re-authorize after a few days, which was frustating.  However, I realized that after publishing I did not delete my token.  So after deleting the token and re-authorizing with the app being "published" (out of testing mode), this seems to have solved the issue.  I am currently on day 14 since then without having to re-authorize again.  Thank you so much for figuring this out!

**Comment by Calcu2006** [comment]
I found that "sometimes" the refresh time of the photos is 2 every second... without touching anything, suddenly my photos are changing 2 or 3 photos each second, and that's why my token expires too fast.

Someone found this too ?

**Comment by aneaville** [comment]
Calcu2006, 
I have noticed this a couple times. I believe this is symptom of the token already being expired, and shouldn't be able to contribute to it expiring any sooner. Please open a new issue and list your findings. Could you also address whether you see this every time the picture changes, or only occasionally.

**Comment by aneaville** [comment]
Well I am on day 10. With a fresh restart this morning, I still haven't needed to re-authorize. I think the above listed method of changing the google project to 'published' is a valid fix for this issue. Anyone want to take notes as they make the change themselves and add it to the installation instructions?

**Comment by Sachin1968** [comment]
Changed it to Pubished and will reauthorize it. I did see a message that says I need to submit the app to Google for Verification though?

Verification Status
 Needs verification

Because you're using one or more sensitive scopes, your app registration requires verification by Google. Please prepare your app to submit for verification.

**Comment by aneaville** [comment]
For any one still following along, I have not had to re-authorize for over 2 months now.

**Comment by Sachin1968** [comment]
> For any one still following along, I have not had to re-authorize for over 2 months now.

Same here. Working fine for a few months now.

**Comment by Andreslion** [comment]
I have never had this kind of issues with my own photo frame. 
That said though, I did another photo frame for my cousin: this one kept getting "disconnected" from google, just freezing the pictures, then showing the dreaded whole black screen with "loading..." at the bottom.
I just found this thread, so I went and compare the OAuth consent screen Publishing status for both projects, and lo and behold, my Publishing status was "In production" (never giving me any trouble), but the other project Status was in "Testing". Sure thing I changed it right away to to production. 
Hope this will do it as it did for you guys. 
Will report back in a few weeks after my cousin won't complain anymore as his frame is not working. XD
